join-domain/windows/init.sls

```
@@ -1,18 +1,31 @@
 {%- from tpldir + '/map.jinja' import join_domain with context %}
 
+manage join domain script:
+  file.managed:
+    - name: {{ join_domain.script.name }}
+    - source: {{ join_domain.script.source }}
+    - makedirs: true
+
 join standalone system to domain:
-  cmd.script:
-    - name: salt://{{ tpldir }}/files/JoinDomain.ps1
-    - args: >-
+  cmd.run:
+    - name: >-
+        powershell "{{ join_domain.script.name }}"
```

The `shell: powershell` line handles the powershell invocation. This just needs to call the script.

```suggestion
        & "{{ join_domain.script.name }}"
```

---

I explicitly put in powershell instead of & as the `&` option will run it in the background and suppress any errors from surfacing.

---

This needs some proof. The `-ErrorAction Stop` argument should force an errexit, for the ps script. Salt already calls this with `Powershell -NonInteractive -NoProfile`. Invoking another powershell process should be wholly unnecessary.

---

here is an example of a success and a forced failure, using `cmd.run` and `& <script>`:

```
C:\WINDOWS\system32> C:\salt\salt-call.bat state.sls foo
local:
----------
          ID: run foo.ps1
    Function: cmd.run
        Name: & "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop
      Result: True
     Comment: Command "& "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop" run
     Started: 18:30:47.222000
    Duration: 425.0 ms
     Changes:
              ----------
              pid:
                  15064
              retcode:
                  0
              stderr:
              stdout:
                  i like cheese

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time: 425.000 ms
C:\WINDOWS\system32> C:\salt\salt-call.bat state.sls foo
[ERROR   ] Command 'chcp 437 > nul & Powershell -NonInteractive -NoProfile "& \"c:/salt/srv/salt/foo/foo.ps1\" -SecureString (ConvertTo-SecureString -String \"i like cheese\" -AsPlainText -Force) -ErrorAction Stop"' failed with return code: 1
[ERROR   ] stderr: ScriptHalted
At C:\salt\srv\salt\foo\foo.ps1:6 char:1
+ throw
+ ~~~~~
    + CategoryInfo          : OperationStopped: (:) [], RuntimeException
    + FullyQualifiedErrorId : ScriptHalted
[ERROR   ] retcode: 1
[ERROR   ] {u'pid': 24320, u'retcode': 1, u'stderr': u'ScriptHalted\r\nAt C:\\salt\\srv\\salt\\foo\\foo.ps1:6 char:1\r\n+ throw\r\n+ ~~~~~\r\n    + CategoryInfo          : OperationStopped: (:) [], RuntimeException\r\n    + FullyQualifiedErrorId : ScriptHalted', u'stdout': u''}
local:
----------
          ID: run foo.ps1
    Function: cmd.run
        Name: & "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop
      Result: False
     Comment: Command "& "c:/salt/srv/salt/foo/foo.ps1" -SecureString (ConvertTo-SecureString -String "i like cheese" -AsPlainText -Force) -ErrorAction Stop" run
     Started: 18:31:16.711000
    Duration: 517.0 ms
     Changes:
              ----------
              pid:
                  24320
              retcode:
                  1
              stderr:
                  ScriptHalted
                  At C:\salt\srv\salt\foo\foo.ps1:6 char:1
                  + throw
                  + ~~~~~
                      + CategoryInfo          : OperationStopped: (:) [], RuntimeException
                      + FullyQualifiedErrorId : ScriptHalted
              stdout:

Summary for local
------------
Succeeded: 0 (changed=1)
Failed:    1
------------
Total states run:     1
Total run time: 517.000 ms
```